### PR TITLE
test: Improve test coverage of SelectCoinsBnB

### PR DESCRIPTION
**Pull request by Empact**
There were a few test cases in SelectCoinsBnB testing that did not properly address the code under test:

* Negative effective value test did not trigger the condition so was redundant with basic value selection tests. Note I switched from `assert` to `throw` in order to make this condition testable.
* early bailout optimization tests did not test the actual selection because their utxo pool was polluted by the `make_hard_case` test preceding.
* None of the tests asserted on the value_ret output value. This output is very relevant to sanity checks and more informative in the case of failures where the returned value has changed, because `equal_sets` does not provide information on its failure.

### CODE REVIEW

File: src/wallet/coinselection.cpp [hunk 1]
```diff
@@ -71,8 +73,10 @@ bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_va
     // Calculate curr_available_value
     CAmount curr_available_value = 0;
     for (const CInputCoin& utxo : utxo_pool) {
-        // Assert that this utxo is not negative. It should never be negative, effective value calculation should have removed it
-        assert(utxo.effective_value > 0);
+        if (utxo.effective_value <= 0) {
+            // It should never be negative, effective value calculation should have removed it
```
Comment (qmma70) [inline src/wallet/coinselection.cpp]: Minor: negative **or zero**

**Review by qmma70 — COMMENTED** [review]
utACK

**Comment by Empact** [comment]
Updated comment

**Comment by Empact** [comment]
Improved effective value test case comments

**Comment by Empact** [comment]
Rebased for #12257

**Pull request closed without merging**